File: examples/runtimes/csharp/src/Nodes/HelloWorldNode.cs
using System.Text.Json;

namespace Blok.Runtime.Nodes;

/// <summary>
/// HelloWorldNode is an example Blok node implemented in C#.
/// It reads a name from the request body and a greeting prefix from the
/// node configuration, then returns a greeting message.
/// </summary>
public class HelloWorldNode : INodeHandler
{
    /// <inheritdoc />
    public Task<JsonElement> ExecuteAsync(Context ctx, Dictionary<string, JsonElement> config)
    {
        // Get name from request body or use default
        var name = "World";

        if (ctx.Request.Body.ValueKind == JsonValueKind.Object
            && ctx.Request.Body.TryGetProperty("name", out var nameElement)
            && nameElement.ValueKind == JsonValueKind.String)
        {
            name = nameElement.GetString() ?? "World";
        }

        // Get greeting prefix from config or use default
        var prefix = "Hello";

        if (config.TryGetValue("prefix", out var prefixElement)
            && prefixElement.ValueKind == JsonValueKind.String)
        {
            prefix = prefixElement.GetString() ?? "Hello";
        }

        var message = $"{prefix}, {name}!";
        var timestamp = DateTimeOffset.UtcNow;

        // Store in context vars for downstream nodes
        ctx.Vars["greeting"] = JsonSerializer.SerializeToElement(message);
        ctx.Vars["timestamp"] = JsonSerializer.SerializeToElement(timestamp.ToUnixTimeSeconds());

        // Build the response
        var response = new Dictionary<string, object>
        {
            ["message"] = message,
            ["timestamp"] = timestamp.ToString("o"),
            ["language"] = "C#"
        };

        var result = JsonSerializer.SerializeToElement(response);
        return Task.FromResult(result);
    }
}

File: examples/runtimes/php/src/Nodes/HelloWorldNode.php
<?php

declare(strict_types=1);

namespace Blok\Nodes;

use Blok\Context;
use Blok\NodeHandler;

/**
 * HelloWorldNode is an example Blok node in PHP.
 *
 * Reads "name" from the request body and "prefix" from node config,
 * then returns a greeting message.
 */
class HelloWorldNode implements NodeHandler
{
    /**
     * Execute the hello world node.
     *
     * @param Context $ctx    The workflow execution context.
     * @param array   $config Node-specific configuration (supports "prefix" key).
     * @return array The result containing message, timestamp, and language.
     */
    public function execute(Context $ctx, array $config): mixed
    {
        // Get name from request body or use default
        $name = 'World';
        if (is_array($ctx->request->body) && isset($ctx->request->body['name'])) {
            $name = (string) $ctx->request->body['name'];
        }

        // Get greeting prefix from config or use default
        $prefix = $config['prefix'] ?? 'Hello';

        $message = "{$prefix}, {$name}!";

        // Store in context vars for downstream nodes
        $ctx->vars['greeting'] = $message;
        $ctx->vars['timestamp'] = time();

        // Return response
        return [
            'message' => $message,
            'timestamp' => date('c'),
            'language' => 'PHP',
        ];
    }
}
